docs(spin): Adds canary cmd docs. (#1557)
spinnaker/spinnaker.github.io: guides/spin/canary-configs/index.md

---
layout: single
title:  "Manage Canary Configs"
sidebar:
  nav: guides
---

{% include toc %}

## Overview

> `spin` can manage the whole lifecycle of your canary configs:

Reviewer: ```suggestion
You can use `spin` to manage the whole lifecycle of your canary configs:
```

```bash
$ spin canary canary-configs
Usage:
   canary canary-config [command]

Aliases:
  canary-config, canary-configs, cc

Available Commands:
  delete      Delete the provided canary config
  get         Get the canary config with the provided id
  list        List the canary configs
  retro       Retro the provided canary config
  save        Save the provided canary config

Flags:
  -h, --help   help for canary-config

Global Flags:
      --config string            path to config file (default $HOME/.spin/config)
      --default-headers string   configure default headers for gate client as comma separated list (e.g. key1=value1,key2=value2)
      --gate-endpoint string     Gate (API server) endpoint (default http://localhost:8084)
  -k, --insecure                 ignore certificate errors
      --no-color                 disable color (default true)
      --output string            configure output formatting
  -q, --quiet                    squelch non-essential output

Use " canary canary-config [command] --help" for more information about a command.
```

> The following assumes Spinnaker is running and Gate is listening on `http://localhost:8084`. If

Reviewer: ```suggestion
The following instructions assume Spinnaker is running and Gate is listening on `http://localhost:8084`. If
```

gate is running elsewhere, you can set the Gate endpoint with the global `--gate-endpoint` flag.

> ## Managing Your Canary Config's Lifecycle

Reviewer: ```suggestion

```

> ### Create and update canary configs with `save`

Reviewer: ```suggestion
## Create and update canary configs with `save`
```

```bash
$ spin canary canary-configs save --file <path to canary config json>
```

> Note that `save` accepts the canary config in JSON format. You can quickly export an existing

Reviewer: IIUC...

Would it work to say
`$ spin canary canary-configs save|get --file <path to canary config json>`

And then the explanation would be
Where:
* `save` accepts the canary config in JSON format from the file identified using the `--file` argument
* `get` exports an existing canary config into the identified file.

Author: Sure

Author: Care to make these in a follow-up? We could also make similar changes in `pipeline-templates`, etc.

Reviewer: Yeah, I can do that.

Author: Just to clarify also:

> get exports an existing canary config into the identified file.

`get` just dumps the json to stdout, but you can save that in a file if you want. There's no explicit output file for `get`.

canary config into a valid argument to the `--file` flag by using the `get` command.

> ### List canary configs with `list`

Reviewer: ```suggestion
## List canary configs with `list`
```

```bash
spin canary canary-config list

[
...
{
  "id": "canaryConfigId",
  "name": "canaryConfigName"
  ...
},
...
]

```

> ### Retrieve a single canary config with `get`

Reviewer: ```suggestion
## Retrieve a single canary config with `get`
```

```bash

> spin canary canary-config get canaryConfigId

Reviewer: * The `save` version above is `canary-configs` (plural). This one is singular
* Is this different from the `get` version mentioned in that same place? That one takes a `--file`, but this one takes `canaryConfigId`, but are we doing the same thing here?

Author: > That one takes a --file

It's the same `get. The `--file` belongs to `save`. That sentence was explaining how to use `get` as an argument to the `--file` param of `save.

{
  "id": "myPipelineTemplate"
  ...
}
```

> ### Delete a canary config with `delete`

Reviewer: ```suggestion
## Delete a canary config with `delete`
```

```bash
spin canary canary-config delete canaryConfigId
```

> ### Test a canary config with `retro`

Reviewer: ```suggestion
## Test a canary config with `retro`
```
